analyses/chromosomal-instability/util/circos-plots.R

```diff
@@ -0,0 +1,338 @@
+      # Make color palette based on 5 colors
+      palette_col <- RColorBrewer::brewer.pal(5, name = palette)
+
+      # Make color ramp function based on quantiles and specified palette
+      color_fun <- circlize::colorRamp2(
+        breaks = quantile(
+          bed_df$color_key,
+          c(0.15, 0.35, 0.5, 0.65, 0.85)
+        ),
+        colors = palette_col
+      )
+      # Set up the color ready data.frame
+      bed_df <- bed_df %>%
+        # Make column that specifies the color for each value
+        dplyr::mutate(color_key = color_fun(value))
+
+      # If color key is not numeric...
+    } else {
+
+      # Try to coerce to factor...
+      if (!is.factor(bed_df$color_key)) {
+        bed_df <- bed_df %>%
+          dplyr::mutate(color_key = as.factor(color_key))
+
+        # Print warning about coercion
+        message("`colour_key` argument is being coerced to a factor.")
+      }
+
+      # Determine the number of levels
+      n_levels <- length(levels(bed_df$color_key))
+
+      # ColorBrewer complains if given a number less than 3
+      # Set up a palette based on number of factor levels
+      palette_col <- RColorBrewer::brewer.pal(ifelse(n_levels < 3, 3, n_levels),
+        name = palette
+      )
+
+      # Make a recode color key list
+      level_key <- palette_col[1:n_levels]
+
+      # Make the palette colors the names in this list
+      names(level_key) <- levels(bed_df$color_key)
+
+      # Set up the color ready data.frame
+      bed_df <- bed_df %>%
+        # Make column that specifies the color for each value
+        dplyr::mutate(
+          color_key = dplyr::recode(color_key, !!!level_key),
+          color_key = as.character(color_key)
+        )
+
+      # Only keep the colors that were actually recoded
+      palette_col <- unique(bed_df$color_key)
+    }
+  } else {
+    # If no color key is specified, just make the color black.
+    palette_col <- "black"
+  }
+
+  # Check if the chromosomes have the `chr` format. If not, add it.
+  if (!grepl("chr", bed_df$chr[1])) {
+    bed_df <- bed_df %>%
+      dplyr::mutate(chr = paste0("chr", chr))
+  }
+
+  # If we aren't adding on a track to a current plot...
+  if (!add_track) {
+    # Initialize the plot
+    circlize::circos.clear()
+    circlize::circos.par(start.degree = 90)
+    circlize::circos.initializeWithIdeogram(species = "hg38")
+  }
+
+  if (type == "point") {
+    # Circlize wants things in their own column for each color
+    bed_df <- bed_df %>%
+      # Make a color column and then spread to each so circlize can use it
+      tidyr::spread(color_key, y_val) %>%
+      dplyr::select(-rowid)
+
+    # Add scatterplot track
+    circlize::circos.genomicTrackPlotRegion(bed_df,
+      track.height = track_height,
+      # Establish the ylim
+      ylim = c(y_min, y_max),
+      panel.fun = function(region, value, ...) {
+        circlize::circos.genomicPoints(region, value,
+          col = palette_col
+        )
+      }
+    )
```

Comment:
I'm fairly confident you tested this, but according to the docs: 
> If there is only one numeric column, the length of col can be either one or number of rows of region

So you should be able to use the **unmodified** bed_df and 
```
circlize::circos.genomicPoints(region, value[[1]],
          col = value[[2]]
)
```

Reply:
I'll try again, but I do remember trying this and it wouldn't work.

Reply:
So this works with `genomicPoints`, but not `genomicLines` which is a pain.

Reply:
I've left the lines in, because I think there may be uses for them, but I don't highlight them in the example notebook.